mesh/mesh.go

```diff
@@ -394,31 +409,43 @@ func (msh *Mesh) getValidBlockIDs(ctx context.Context, layerID types.LayerID) ([
 }
 
 // apply the state of a range of layers, including re-adding transactions from invalid blocks to the mempool.
-func (msh *Mesh) pushLayersToState(ctx context.Context, oldVerified, newVerified types.LayerID) {
+func (msh *Mesh) pushLayersToState(ctx context.Context, from, to types.LayerID) error {
 	logger := msh.WithContext(ctx).WithFields(
-		log.FieldNamed("old_verified", oldVerified),
-		log.FieldNamed("new_verified", newVerified))
+		log.Stringer("from_layer", from),
+		log.Stringer("to_layer", to))
 	logger.Info("pushing layers to state")
-	if oldVerified.Before(types.GetEffectiveGenesis()) || newVerified.Before(types.GetEffectiveGenesis()) {
+	if from.Before(types.GetEffectiveGenesis()) || to.Before(types.GetEffectiveGenesis()) {
 		logger.Panic("tried to push genesis layers")
 	}
-
 	// we never reapply the state of oldVerified. note that state reversions must be handled separately.
-	for layerID := oldVerified.Add(1); !layerID.After(newVerified); layerID = layerID.Add(1) {
-		l, err := msh.GetLayer(layerID)
-		// TODO: propagate/handle error
-		if err != nil || l == nil {
-			logger.With().Error("failed to get layer", layerID, log.Err(err))
-			return
+	for lid := from.Add(1); !lid.After(to); lid = lid.Add(1) {
+		if err := msh.pushLayer(ctx, logger, lid); err != nil {
+			logger.With().Warning("failed to push layer. reverting processed layer",
+				log.Err(err),
+				lid,
+			)
+			msh.overwriteProcessed(lid.Sub(1))
```

Comment:
i find this hacky and confusing. 
in a steady-state, processedLayer (verified+1) is in lockstep with tortoise verified layer.
now for this rare occasion, processed is the older than tortoise verified layer.

i think we should change syncer instead of the logic in mesh. maybe syncer should be based off applied layer in SVM rather than processed layer?

Reply:
i think that doesn't work. sync needs to make progress otherwise we may not have enough votes to verify the layer in tortoise and therefore applied layer in svm will be stuck

Reply:
> we may not have enough votes to verify the layer in tortoise and therefore applied layer in svm will be stuck

applied layer in svm will be stuck anyway if any TX is not available. i don't see advancing tortoise verified layer help the situation. what's the point of tortoise making progress if applied layer is stuck?

in any case, i think you can make syncer starting to sync from appliedLayer+1, and still advance processed layer.
appliedLayer(svm) can be completely separate from processedLayer(mesh) and verifiedLayer(tortoise). you can advance the latter two with the appliedLayer stuck.

Reply:
i also don't think we should rush this change. what can't we remove the panic stmt for now and tolerate inconsistent state in the network?

Reply:
> i also don't think we should rush this change. what can't we remove the panic stmt for now and tolerate inconsistent state in the network?

i think this change is perfectly fine. also i don't understand what you are suggesting

Reply:
i think what you suggested leads to redundant work. the verified layer may be stuck but all the data is available, but sync will loop through layers (because it will start from layer in state) and download data from disk.

it is hard for me to say how impactful that work will be, but given that we are using only 2 vcpus on cloud nodes, it can consume 1 of them fully.

Reply:
> i also don't think we should rush this change. what can't we remove the panic stmt for now and tolerate inconsistent state in the network?

i thought about it and i guess what you is that it is ok to submit a layer without some transactions. i don't think that this is ok, because basically it means that after several days every node in the network will have different state. at this point we may even disable transactions and run only consensus because state machine is meaningless if no one can agree on what the state looks like.

so to summarize:

>  i think you can make syncer starting to sync from appliedLayer+1, and still advance processed layer.
appliedLayer(svm) can be completely separate from processedLayer(mesh) and verifiedLayer(tortoise). you can advance the latter two with the appliedLayer stuck.

this doesn't work because of the redundant work. solution in this change does more work only when there is a lack of data

> what can't we remove the panic stmt for now and tolerate inconsistent state in the network?

this doesn't work because this is not a rare issue. after a day every node hit it at least once, and according to panics not one transaction was missing but many. and anyway this is a temporary hack, i don't know what would be the reason to implement it when the solution to this problem is very clear

Reply:
> this doesn't work because of the redundant work. solution in this change does more work only when there is a lack of data

it is not too complicated to add code to syncer that it directly try to fetch contents in valid blocks fro appliedLayer+1 to processedLayer if applied layer is stuck. and with fetcher its a noop if the block data is available locally. syncer just need to be exposed to more APIs from mesh and fetcher. if you want, i can work on that part in this PR.

my point about the applied layer being independent of processedLayer/verifiedLayer is that the node should be able to make consensus progress (advancing verifiedLayer) without ever applying the TXs in any block to SVM. i.e. all TXs stay in pending state. 

i like this PR in that it leverage syncer as a backup when data is missing. the part i dont like is to revert processedLayer only because some TXs are missing and therefore cannot be applied to SVM state. the logic doesnt follow because like i stated above, they should be independent. consensus outcome should not be impacted by whether we are able to apply to the state.

additionally, it makes the mesh code more complex. processed layer can be reverted randomly. i strongly prefer the processed layer only progress in one monolithic direction and in order. it will make future debugging a lot easier.

Reply:
> and with fetcher its a noop if the block data is available locally

it is not noop. it will read data from disk in a loop even when applied layer is legitimately stuck (not due to the missing data)

Reply:
>  i strongly prefer the processed layer only to progress in one monolithic direction and in order. it will make future debugging a lot easier.

we can then have a queue of layers that failed to be applied to the state, new layers will be enqueued where i am overwriting processed layer. and dequeued when sync round starts, does it work?

it will do even less redundant work

Reply:
we just need to add HasTX API in the db and read the keys. i think this overhead is worth it if processedLayer can be kept simple.

Reply:
besides, i think we also need to discuss data availability in general. not just missing TXs.

Reply:
> we just need to add HasTX API in the db and read the keys. i think this overhead is worth it if processedLayer can be kept simple.

i realized that we don't need it to be in db. we can have a map in mesh that will keep layers with missing data. layer will be added to the map only if mesh failed to process the layer. so it is kind of a queue.

syncer will read that map and for layer between applied + 1 ... processed it will fetch data if the key is in the map.

> besides, i think we also need to discuss data availability in general. not just missing TXs.

sure, we can discuss it. in my understanding, it is always the same. syncer should be the best effort until we know that the data is absolutely required (after tortoise verified layer). and it will work generically both if the ballot was pruned or data is being maliciously withhold

Reply:
> i realized that we don't need it to be in db. we can have a map in mesh that will keep layers with missing data. layer will be added to the map only if mesh failed to process the layer. so it is kind of a queue.

syncer will read that map and for layer between applied + 1 ... processed it will fetch data if the key is in the map.

i am happy w this solution. this means the logic around processedLayer stay the same.

> sure, we can discuss it. in my understanding, it is always the same. syncer should be the best effort until we know that the data is absolutely required (after tortoise verified layer). and it will work generically both if the ballot was pruned or data is being maliciously withhold

yes. id like to discuss the general strategy in the dev meeting on Wednesday first and see if this PR fit in the picture